### DISCUSSION TITLE
Rewriting the internals of Leon

### DISCUSSION THREAD

**Discussion by svenjacobs** [leon-cleaning-services/leon#315] [Announcements]
I'm currently rewriting the internals of Leon. Instead of sanitizers based on regular expressions, the idea is that Leon creates an internal representation of an address that contains the domain, path and query parameters. Sanitizers can then declare which parameters should be removed. This change will also allow the UI to show all parameters so that users can manually select or deselect parameters for removal. Hence Leon will also work with "unknown" addresses.

Until this refactoring is completed, I will not add any new sanitizers because they need to be rewritten anyway.

**Comment by serrq** [comment]
Okay. For Christmas is ready the new Leon?

**Reply by svenjacobs** [reply]
I don't know how long it takes. It's done when it's done 😉

**Reply by serrq** [reply]
Not a tentative to hurry up. Just to realize.

**Reply by rsb-23** [reply]
I just looked at **Url.kt** from the refactoring branch. Can we use `import java.net.URL` or `import android.net.Uri` url parser instead?
Similar to [decoding](https://github.com/svenjacobs/leon/blob/927416199f3db0eebf581ba5b544be3c792c4fbc/core-common/src/main/kotlin/com/svenjacobs/app/leon/core/common/url/DecodeUrl.kt#L21) done in **DecodeUrl.kt**

**Reply by svenjacobs** [reply]
> I just looked at **Url.kt** from the refactoring branch. Can we use `import java.net.URL` or `import android.net.Uri` url parser instead? Similar to decoding done in **DecodeUrl.kt**

I wanted to have a plain Kotlin class. The `common` module is a plain Kotlin module, no Android. So we can't use `Uri` from Android here. What would be the benefits of using `java.net.URL`?

**Reply by rsb-23** [reply]
- In general, regex are slower than other approach.
- Why to reinvent the wheel, if existing library works and fits in architecture.
- Less code & error -> Less testing -> Easy maintainability.

**Reply by svenjacobs** [reply]
I agree with your points except

> if existing library works and fits in architecture

I try to keep the common code compatible to Kotlin Multiplatform. Who knows, maybe in the future we have a multiplatform cleaning library?

**Reply by TPS** [reply]
> Can we use `import java.net.URL` or `import android.net.Uri` url parser instead?

https://github.com/ethauvin/urlencoder#why-not-simply-use-javaneturlencoder denotes some errors in those, also.

**Reply by svenjacobs** [reply]
> > Can we use `import java.net.URL` or `import android.net.Uri` url parser instead?
> 
> https://github.com/ethauvin/urlencoder denotes some errors in those, also.

Thanks for the link. That's an interesting alternative for `URLDecoder`.

Regarding `Url.kt`, the intention was to have a class that splits an URL into its components so that these components can than be used by cleaning rules.

**Reply by rsb-23** [reply]
Maybe there's a confusion, l am not referring to Url encoder/decoder. 
Hence rephrasing my query :
- `java.net.URL` splits an url into its components. (what Url.kt is intended)
- Since `java.net.URLDecoder` is already used in project, doesn't it mean `java.net.URL` is also compatible to Kotlin Multiplatform?
- Shared Url _encoder_ project does not mention any issue with the url **parser**.

**Reply by TPS** [reply]
> - Shared Url _encoder_ project does not mention any issue with the url **parser**. 

@rsb-23 Maybe I misunderstand your details, but here's what the relevant section at the above repo link says *(emphasis mine)*:
>For example, a simple search query such as:  
```kotlin  
val u = URLEncoder.encode("foo +bar", StandardCharsets.UTF_8)  
```  
>would be encoded as:  
```  
foo+%2Bbar  
```  
>Trying to decode it with [Spring](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/util/UriUtils.html#decode(java.lang.String,java.lang.String)), for example:  
```kotlin  
UriUtils.decode(u,   StandardCharsets.UTF_8)  
```  
>would return:  
```  
foo++bar  
```  
>_**Unfortunately, decoding with [Uri.decode](https://developer.android.com/reference/android/net/Uri#decode(java.lang.String)) on Android, [decodeURI](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/decodeURI) in Javascript, etc. would yield the exact same result.**_  
![URLEncoder](https://live.staticflickr.com/65535/52607534147_6197b42666_z.jpg)

**Reply by rsb-23** [reply]
@TPS I had understood that the UrlEncoder.encode is not consistent with decode logic.

But my query is about **parser**. 
> Parser and encoder are different functionality, please don't mix the two.

**Comment by farOverNinethousand** [comment]
Hi Sven,
I'm currently writing a Telegram Bot which will have similar functionality to your application (if something like this already exists please let me know but I didn't find anything nice and open source).
While a native app like yours is more powerful, I will still work on releasing said bot.
During my research about similar tools I found yours to be nearly the only actively developed project of that kind.

Now I'm asking myself:
Will you design the new link-sanatizer-system in a way that it can be programmatically used in other applications?
I'm asking as this could save double-work and I'd love to add URL-sanatizers to my bot which will also be usable in other applications :)

Thanks for your time!

**Reply by svenjacobs** [reply]
Hello @farOverNinethousand,

thanks for your feedback. Currently I did not plan to write this as a library or sorts of which can be used by other applications but that's an interesting idea. I'm not familiar with Telegram Bot development. What language do you write Bots in and how would you be able to call other applications?

**Reply by farOverNinethousand** [reply]
I mainly ment the separation between your app and the link cleaning functionality.
In this case the bot is written in Python (because there is a nice bot library available in Python) and the lib is also written in Python.
My motivation for the bot was actually only to avoid those mydealz tracking URLs, then I found your project.
Here is the work I did this weekend:
https://github.com/farOverNinethousand/URL-Tracking-remover-Telegram-bot
All URL cleaning functionality is separated from the bot class and can be used within other projects:
https://github.com/farOverNinethousand/URL-Tracking-remover-Telegram-bot/blob/main/URLCleaner.py
The "Cleaning Rules" aren't yet exportable yet but from what is there now it would only be a work of like one hour to make those rules exportable/importable as json format.
You can see that I re-built a lot of your "Cleaning Rules" in Python. For testing I also added some which were suggested by other users in issues in your project.
I will add a live version of said bot some time this week though I have to admit your project is way more useful, at least on Android.
EDIT Screenshot:
![url_cleaner_bot_showcase](https://github.com/svenjacobs/leon/assets/36418572/7b60e382-bac7-4855-8427-dca919d71b38)

**Reply by svenjacobs** [reply]
> I mainly ment the separation between your app and the link cleaning functionality.

I understand. The link cleaning functionality could be extracted into a library. But it would be a Kotlin library. How useful is this for a Telegram bot? 🤔

**Reply by farOverNinethousand** [reply]
> How useful is this for a Telegram bot?

The Kotlin library would not be helpful for me but can still be helpful for other projects and [writing Telegram bots in Kotlin](https://github.com/kotlin-telegram-bot/kotlin-telegram-bot) is also possible.
In the end it doesn't really matter:
My main point is: The functionality you provide via Leon is useful for other projects too so writing those rules only once and having the ability to re-use them in other applications would be nice.

Maybe I overlook something - which points mitilate against my idea?

**Reply by svenjacobs** [reply]
> Maybe I overlook something - which points mitilate against my idea?

Nothing, I think this is a good idea. I just want to understand your use case 😃

**Reply by yvolk** [reply]
I want to follow up on the idea of having this app's functionality as a reusable thing, e.g. as a Kotlin library.
This could be a library that can be used by any other Java/Kotlin app (and not only in Android!) AND the same library could be used by this "Leon" app! So even @svenjacobs the Author of the URL cleaning tool wouldn't have to write the same things twice.
Additional bonus: ability to use the library outside of Android world. You only need to write the "URL cleaning library" without any Android-specific dependencies. Ideally - without any dependencies, so it could be easily injected everywhere.

PS: Hi @svenjacobs, I'm the developer of AndStatus Android app https://github.com/andstatus/andstatus
See your comment https://github.com/andstatus/andstatus/issues/545#issuecomment-2148935330

**Reply by svenjacobs** [reply]
Hello @yvolk, thanks for your comment!
Yes this would be a really cool thing. Unfortunately I didn't have time yet to work on a Leon library.

**Reply by yvolk** [reply]
No problem! We're not in a hurry :-)
Please let me know when your library is ready for testing.

**Reply by farOverNinethousand** [reply]
@yvolk I have written such a library as part of [another project](https://github.com/farOverNinethousand/URL-Tracking-remover-Telegram-bot/blob/main/URLCleaner.py).
This is not advertising - just saying that it can be used as an inspiration.

**Reply by yvolk** [reply]
@farOverNinethousand Looks like what we really need as a minimum are that cleaning rules in some platform-independent format that could be maintained in one place, and downloaded e.g. during app build/update. Looking at your source code it's easy to imagine corresponding JSON with rules definitions :-)

**Reply by svenjacobs** [reply]
@yvolk @farOverNinethousand Maybe we don't need to reinvent the wheel? Maybe we could use already existing databases like from [ClearURLs](https://docs.clearurls.xyz/1.26.1/specs/rules/) for example?

**Reply by farOverNinethousand** [reply]
@svenjacobs Looks cool and looks like it is still actively maintained :)

**Reply by yvolk** [reply]
> Maybe we don't need to reinvent the wheel? Maybe we could use already existing databases like from [ClearURLs](https://docs.clearurls.xyz/1.26.1/specs/rules/) for example?

Do you mean Leon could start using it too instead of current rules list that you maintain?!

**Reply by svenjacobs** [reply]
> Do you mean Leon could start using it too instead of current rules list that you maintain?!

Yes, or in addition to it.

**Reply by farOverNinethousand** [reply]
Soo I've looked into the ClearURLs project and also played around a bit.
Firstly, the ClearURLs project is currently [undergoing a rewrite](https://github.com/ClearURLs/Addon/issues/74) (ahh since 2020?) so I do not expect any one of them to find time for the library idea.
Also, same as Leon and all other projects like this, it is lacking the ability for users to [create custom rules](https://github.com/ClearURLs/Addon/pull/250) which would make it so much powerful.

Then as for my own project-idea, that is written in Python so I'm unsure whether a Python library would be helpful for things like a browser addon though in th end, even if such projects wouldn't share the same library, the ability to use exported rules across multiple of such rules is what I would consider the main goal (so I agree with you, @yvolk).

For testing, I've added import/export functionality to my test-project so all rules of Leon + ClearURLs in my format would look like this:
[cleaningrules_save_test.json](https://github.com/user-attachments/files/15936971/cleaningrules_save_test.json)
--> Total of 231 rules, already usable.

I've found weaknesses in both mine and the rules of ClearURLs though I was able to get all of them working in one project.
Even if we had a perfectly working library, we'd still need to find a way to merge all rules of such projects, maybe a serverside application/DB would be needed for this.

**Reply by svenjacobs** [reply]
> Firstly, the ClearURLs project is currently undergoing a rewrite (ahh since 2020?) so I do not expect any one of them to find time for the library idea.

My idea was that Leon (or the library) can read the ClearURLs configuration. I don't think we need their help for this. These files are publicly hosted on GitHub / GitLab and could be downloaded regularly. Either Leon downloads and interprets them on the fly or we write an application (an importer) that converts the rules into a common format, maybe even Kotlin directly? These rules could then be bundled with regular releases of Leon / the library?

**Reply by farOverNinethousand** [reply]
Yap good idea.
As long as the file is hosted on github, Leon could just just download it periodically.
I still think it would be interesting to get some input from the ClearURLs devs - maybe they got some good ideas for an interesting rule structure for their rewrite.

**Reply by svenjacobs** [reply]
> I still think it would be interesting to get some input from the ClearURLs devs - maybe they got some good ideas for an interesting rule structure for their rewrite.

Yes, absolutely! Maybe we could even define some kind of standard (JSON) file format / structure for cleaning rules? This would require writing some RFC-style technical documentation and getting feedback from everybody.

**Reply by TPS** [reply]
> As long as the file is hosted on github, Leon could just just download it periodically.

@svenjacobs Just a note, Leon currently operates offline-only, so this would a pretty significant permissions-shift (reminds of https://github.com/svenjacobs/leon/discussions/80).  However, I don't see that database updating more often than Leon itself, so maybe this could be incorporated into Leon's own static database to be updated on its release cycle?

**Reply by svenjacobs** [reply]
> > As long as the file is hosted on github, Leon could just just download it periodically.
> 
> @svenjacobs Just a note, Leonb currently operates offline-only, so this would a pretty significant permissions-shift (reminds of https://github.com/svenjacobs/leon/discussions/80).  However, I don't see that database updating more often than Leon itself, so maybe this could be incorporated into Leon's own static database to be updated on its release cycle?

Yes, I'm aware of this. That's why I also mentioned an importer and generating Kotlin code. I think I prefer this solution anyway.

**Reply by TPS** [reply]
> Maybe we could even define some kind of standard (JSON) file format / structure for cleaning rules?

@svenjacobs Would supercede https://github.com/svenjacobs/leon/discussions/158?

**Reply by TPS** [reply]
> Yes, I'm aware of this. That's why I also mentioned an importer and generating Kotlin code. I think I prefer this solution anyway.

@svenjacobs This would also give you some necessary editorial powers, so we users aren't held hostage to an external entity. E.g., d/t @ClearURLs' slower process, the [issues](https://github.com/ClearURLs/Rules/issues) piling up there could be better handled by you directly here, &, whenever they get around to releasing that long-awaited rewrite (can't tell if it's just for the browser addon or for the database also), you could re-evaluate how that works for you.

**Reply by farOverNinethousand** [reply]
Yeah needing less permissions is always nice.
You could even do both, provide the latest set of ClearURLs rules via each app update but also add the ability for users to import those rules, maybe by additionally linking the json URL inside your app.
Remember that users are lazy and most of them will use only the rules Leon is shipping.

**Reply by krystian3w** [reply]
Check:

https://github.com/DandelionSprout/adfilt/blob/master/ClearURLs%20for%20uBo/clear_urls_uboified.txt
https://raw.githubusercontent.com/DandelionSprout/adfilt/master/ClearURLs%20for%20uBo/clear_urls_uboified.txt

**Reply by TPS** [reply]
Also see the discussions @ https://github.com/DandelionSprout/adfilt/discussions/163 (as linked inside the above) & https://github.com/AdguardTeam/FiltersRegistry/issues/962 (where it's noted that https://github.com/DandelionSprout/adfilt/blob/master/ClearURLs%20for%20uBo%2Fcompile.py converts ClearURLs → adblock rules, but might be adapted, [like above](https://github.com/svenjacobs/leon/discussions/315#discussioncomment-9847206), to other formats [as required](https://github.com/svenjacobs/leon/discussions/315#discussioncomment-9851690)).

**Reply by svenjacobs** [reply]
> Would supercede https://github.com/svenjacobs/leon/discussions/158?

@TPS Yes, I think if we aim for a general format or even a standard, JSON would be more appropriate. Although it is more verbose it is also more more common and there should be parsers for any language.

**Reply by TPS** [reply]
> Yes, I think if we aim for a general format or even a standard, JSON would be more appropriate.

Here's perhaps a devil's-advocate perspective to consider:  why invent [another new format](https://xkcd.com/927/) when the current adblock syntax _might_ do fine?  It's very mature, virtually universal, extensively extensible, & gives immediate access to [a vast amount of data](https://filterlists.com/) that's a superset of what you're looking for.

**Reply by svenjacobs** [reply]
> > Yes, I think if we aim for a general format or even a standard, JSON would be more appropriate.
> 
> Here's perhaps a devil's-advocate perspective to consider: why invent [another new format](https://xkcd.com/927/) when the current adblock syntax _might_ do fine? It's very mature, virtually universal, extensively extensible, & gives immediate access to [a vast amount of data](https://filterlists.com/) that's a superset of what you're looking for.

Yes sure, sounds good. I'm not yet familiar with the AdBlock/EasyList syntax, which is documented [here](https://adguard.com/kb/de/general/ad-filtering/create-own-filters/). Since the format was mainly created to block requests and not remove parameters from URLs, it might lack some features? [Removing parameters](https://adguard.com/kb/de/general/ad-filtering/create-own-filters/#removeparam-modifier) is supported but what about some more special cases where we rewrite an URL (see [AmazonProductSanitizer](https://github.com/svenjacobs/leon/blob/53a010945583053f10f3df0e67e10da4e6057f04/core-domain/src/main/kotlin/com/svenjacobs/app/leon/core/domain/sanitizer/amazon/AmazonProductSanitizer.kt))? Of course we could argue that Leon loses some features but therefor is able to parse popular block lists?

**Reply by farOverNinethousand** [reply]
@svenjacobs 
Why would Leon lose features?
You could support both, those rules and your own rules.
It shouldn't take much more work.

**Reply by svenjacobs** [reply]
Hm, I'll have to think about it. I would prefer if we go one route: Let's say import all rules from adblock lists and get rid of custom managed sanitizers 🤷🏼

**Reply by TPS** [reply]
> [Removing parameters](https://adguard.com/kb/de/general/ad-filtering/create-own-filters/#removeparam-modifier) is supported but what about some more special cases where we rewrite an URL (see [AmazonProductSanitizer](https://github.com/svenjacobs/leon/blob/53a010945583053f10f3df0e67e10da4e6057f04/core-domain/src/main/kotlin/com/svenjacobs/app/leon/core/domain/sanitizer/amazon/AmazonProductSanitizer.kt))? Of course we could argue that Leon loses some features but therefor is able to parse popular block lists?

[See `$replace` & `$rewrite` on that same writeup](https://adguard.com/kb/general/ad-filtering/create-own-filters/).  This is already added via commonly supported extensions from the [base `AdBlock` syntax](https://help.adblockplus.org/hc/en-us/articles/360062733293-How-to-write-filters).

**Reply by TPS** [reply]
@svenjacobs Just curious as to whether you'd come to any decision re: what direction you're going w/ all this (such as https://github.com/svenjacobs/leon/discussions/158).  No pressure @ all; I see continued development under the _status quo_, & was curious where I could witness any development toward Léon TNG. 🙇🏾‍♂️

**Reply by svenjacobs** [reply]
I'm sorry but I was quite busy the last year and didn't have time to think about this any further.

**Reply by TPS** [reply]
@svenjacobs Again, no worries or apology needed.  We (or @ least I) plan to be here whenever you do pick this back up.

**Reply by svenjacobs** [reply]
So I thought about this approach. I think we should create a Kotlin Multiplatform library and command line application that parses a [JSON file](https://github.com/svenjacobs/leon/discussions/158#discussioncomment-12573395) with sanitizer configurations and performs cleaning and transformation of URLs. Once the library and cmd line application is working, we can integrate it into Leon. The cmd line application can also be used in the build pipeline of Leon to ensure that the JSON file, which then is the only source of sanitizer configurations, is valid.

So this will be a combination of both ideas, rewriting the internals of Leon and having a ~~YAML~~ configuration file.

Once this is done we can think of writing import tools that read other URL cleaning configuration files (ClearURLs for example) and convert them into our format.

FYI @rsb-23

**Reply by svenjacobs** [reply]
As you may have noticed I moved this project to a new organization and also created a skeleton project for the [library](https://github.com/leon-cleaning-services/mathilda).

**Reply by TPS** [reply]
> new organization
> skeleton project

@svenjacobs _Great_ names!  Inspired by people you know?

**Reply by TPS** [reply]
@svenjacobs Now that you've settled on a container format, a thought: 

JSON is designed to be simple to work w/, but trade-off is that it can be very space-hungry, especially since new URL transforms will be static w/ APK updates.  Typical publish toolchain will optimize for APK size for _transmission_, but not once _installed_, which might get substantial once including JSON logic & going from (currently) 10s of transforms to (potentially) 100s or 1000s (especially once you start processing external databases).  Part of Léon's charm has always been that it's < 10MB installed.  What will these upgrades do for that, I wonder?

Perhaps something like https://github.com/shibasis0801/kmm_zlib_compression & linked article will be helpful in your initial design phase?  & taking care to use as much is natively available so as to not bloat that installed size w/ add'l libs to carry around?

**Reply by svenjacobs** [reply]
> > new organization
> > skeleton project
> 
> @svenjacobs _Great_ names!  Inspired by people you know?

I'm not sure if you're joking or don't know the reference 😉 Names are based on the movie [Léon The Professional](https://www.rottentomatoes.com/m/leon_the_professional) 😅 If you haven't seen it, do it!

**Reply by svenjacobs** [reply]
> @svenjacobs Now that you've settled on a container format, a thought: 
> 
> JSON is designed to be simple to work w/, but trade-off is that it can be very space-hungry, especially since new URL transforms will be static w/ APK updates.  Typical publish toolchain will optimize for APK size for _transmission_, but not once _installed_, which might get substantial once including JSON logic & going from (currently) 10s of transforms to (potentially) 100s or 1000s (especially once you start processing external databases).  Part of Léon's charm has always been that it's < 10MB installed.  What will these upgrades do for that, I wonder?

Good point! We could also think about generating Kotlin code from JSON? However I don't know if it will actually reduce the size compared to JSON if we have hundreds or thousands of sanitizers? I guess we have to find out once we're there.

**Comment by TPS** [comment]
@svenjacobs Is https://github.com/leon-cleaning-services/leon/pull/797 the rewrite done?!  I see LLM(s) were used, but what all did you decide re: all the previous discussion?

**Comment by svenjacobs** [comment]
> @svenjacobs Is https://github.com/leon-cleaning-services/leon/pull/797 the rewrite done?!  I see LLM(s) were used, but what all did you decide re: all the previous discussion?

For now I implemented everything from my initial post. So I'd say yes, the rewrite is complete.
